File: internal/engine/driver/browse.go
package driver

import (
	"context"
	"encoding/json"

	"github.com/marlexladag/lantern/internal/engine/dberr"
)

// MaxBrowseLimit caps a single page. An unbounded limit is how a UI bug
// becomes an out-of-memory crash on a table with a hundred million rows.
const MaxBrowseLimit = 1000

// SortKey is one ORDER BY term.
type SortKey struct {
	Column string `json:"column"`
	Desc   bool   `json:"desc"`
}

// BrowseRequest asks for one page of a table, semantically. The caller never
// writes SQL: each driver renders this in its own dialect, which is what
// makes a second engine an added implementation rather than a rewrite.
//
// Browsing is stateless by design: there is no server-side cursor to hold
// open or leak when a UI tab closes. Each request is self-contained, carrying
// whatever the previous page's Keyset handed back in After. That is narrower
// than spec section 7's windowed cursor, which exists for arbitrary SQL — a
// result set that cannot be re-derived and must therefore be held open. A
// table can always be re-derived from its BrowseRequest, so it never needs
// that cursor's cost.
type BrowseRequest struct {
	Database string `json:"database"`
	Table    string `json:"table"`
	// Sort is empty for the driver's stable default, which should be the
	// primary key so that pagination is deterministic. A request may carry a
	// non-empty After alongside an empty Sort: that is the ordinary shape of
	// a second page when the caller never chose a sort, not a contradiction
	// — the driver applies the same default order both times, and the caller
	// is never expected to know what that default is, only to echo back the
	// Keyset it was handed.
	Sort []SortKey `json:"sort,omitempty"`
	// After is the previous page's Keyset, echoed back untouched. Nil for the
	// first page.
	After []Value `json:"after,omitempty"`
	// Offset is used only when the driver told the caller it could not
	// paginate by key.
	Offset int `json:"offset,omitempty"`
	Limit  int `json:"limit"`
}

// Validate rejects a request no driver should be asked to run.
func (r BrowseRequest) Validate() error {
	if r.Table == "" {
		return dberr.New(dberr.KindInvalid, "browse: table is required")
	}
	if r.Limit <= 0 {
		return dberr.New(dberr.KindInvalid, "browse: limit must be positive")
	}
	if r.Limit > MaxBrowseLimit {
		return dberr.New(dberr.KindInvalid, "browse: limit exceeds the maximum page size")
	}
	return nil
}

// BrowsePage is one page of rows.
type BrowsePage struct {
	Columns []ColumnMeta `json:"columns"`
	Rows    [][]Value    `json:"rows"`
	// Keyset carries whatever the driver needs to fetch the next page, taken
	// from the last row. The caller echoes it back in After without
	// understanding it. Absent when the driver paginated by offset, or when
	// the result is exhausted.
	Keyset []Value `json:"keyset,omitempty"`
	// Exhausted is true when fewer rows than Limit came back, so the caller
	// can stop asking.
	Exhausted bool `json:"exhausted"`
	// Offset is the offset of the row after this page, for drivers that could
	// not paginate by key.
	Offset int `json:"offset"`
}

// MarshalJSON guarantees Rows serializes as [] rather than null. Rows is a
// plain slice with no omitempty, so its zero value — nil, exactly what an
// empty table or an exhausted After produces — would otherwise marshal to
// the JSON literal null through encoding/json's default behavior. The
// TypeScript side declares Rows as an array; null landing where an array is
// expected doesn't fail loudly on this one field, it has blanked the whole
// app before. Guaranteed once here rather than trusted to every driver that
// builds a page.
func (p BrowsePage) MarshalJSON() ([]byte, error) {
	type alias BrowsePage
	a := alias(p)
	if a.Rows == nil {
		a.Rows = [][]Value{}
	}
	return json.Marshal(a)
}

// Browser is an OPTIONAL interface. A driver implements it when it can page a
// table; the caller discovers it by type assertion, exactly as spec section 4
// requires, so a schemaless engine is never forced to stub it.
type Browser interface {
	Browse(ctx context.Context, req BrowseRequest) (*BrowsePage, error)
}

File: src/lib/browse.ts
/**
 * Typed client for the engine's table-browsing method.
 *
 * Everything funnels through `request` in ./engine, which is the single
 * `engine_request` command. Do not add a second channel.
 */
import { request } from './engine';

/**
 * One cell's engine-neutral kind, mirroring `driver.ValueKind` in
 * internal/engine/driver/value.go — what the grid branches on: numbers
 * right-align, NULL renders italic and dim, bytes refuse to render inline.
 * browse.test.ts checks this list against the Go one by parsing both,
 * rather than by eye, the same way connections.test.ts already does for
 * DbErrorKind.
 */
export type ValueKind = 'null' | 'text' | 'int' | 'float' | 'bool' | 'bytes' | 'time';

/**
 * One cell, in engine-neutral terms, mirroring `driver.Value`.
 *
 * `text` is always the display form — never a typed payload — because a
 * JSON number is a float64 and would silently destroy an int64 or a
 * DECIMAL's precision above 2^53. `kind` is what lets the grid tell a real
 * NULL apart from the empty string, and the text "null" apart from both.
 */
export interface Value {
  kind: ValueKind;
  text: string;
}

/** One ORDER BY term, mirroring `driver.SortKey`. */
export interface SortKey {
  column: string;
  desc: boolean;
}

/** One column of a result set, mirroring `driver.ColumnMeta`. */
export interface ColumnMeta {
  name: string;
  data_type: string;
}

/**
 * The engine's cap on a single page, mirroring `driver.MaxBrowseLimit`. An
 * unbounded limit is how a UI bug becomes an out-of-memory crash on a table
 * with a hundred million rows.
 */
export const MAX_BROWSE_LIMIT = 1000;

/**
 * One page of rows, exactly as `browse.page` returns it — mirrors
 * `driver.BrowsePage`.
 *
 * `sort_token` is issued alongside `keyset` and names — opaquely, this
 * client never decodes it — the sort that produced it. It stays on this
 * type, in full, rather than being stripped out for the UI: {@link
 * browsePage}'s `after` parameter takes this whole object back, precisely
 * so a caller continuing a page is never handed `keyset` and `sort_token`
 * as two separate values it could assemble incorrectly. The engine refuses
 * a keyset replayed under a mismatched sort_token — that pairing is what
 * turned eight Go tests red the day it became mandatory, every one of them
 * because a test helper built a continuation from the keyset alone. Both
 * are absent exactly when the driver paginated by offset instead, or when
 * the page is exhausted.
 */
export interface BrowsePage {
  columns: ColumnMeta[];
  rows: Value[][];
  keyset?: Value[];
  sort_token?: string;
  exhausted: boolean;
  offset: number;
}

/**
 * Asks for one page of a table, semantically — no SQL, mirroring
 * `driver.BrowseRequest`.
 *
 * `sort` is empty for the driver's stable default (its primary key); a
 * request may carry a non-empty `after` alongside an empty `sort` — that is
 * the ordinary shape of a second page when the caller never chose a sort,
 * not a contradiction, since the driver applies the same default sort both
 * times.
 *
 * `after` is the previous page in full, not a bare keyset array — see
 * {@link BrowsePage}'s own doc comment for why. Omit it for the first page.
 *
 * `offset` only matters to a driver that told the caller it could not
 * paginate by key, in which case the previous page's own `offset` is
 * echoed back directly: unlike a keyset, a bare offset carries no paired
 * secret a caller could separate it from.
 */
export interface BrowseRequest {
  database: string;
  table: string;
  sort?: SortKey[];
  after?: BrowsePage;
  offset?: number;
  limit: number;
}

/** Fetches one page of a table's rows through `browse.page`. */
export function browsePage(sessionId: string, req: BrowseRequest): Promise<BrowsePage> {
  const { after, ...rest } = req;
  return request<BrowsePage>('browse.page', {
    session_id: sessionId,
    ...rest,
    after: after?.keyset,
    sort_token: after?.sort_token,
  });
}
